File: presentation/src/main/java/com/hnh/presentation/logic/classification/EMASmoothing.kt
package com.hnh.presentation.logic.classification

import java.util.*
import java.util.concurrent.LinkedBlockingDeque


/**
 * Runs EMA smoothing over a window with given stream of pose classification results.
 */
class EMASmoothing @JvmOverloads constructor(
    private val windowSize: Int = DEFAULT_WINDOW_SIZE,
    private val alpha: Float = DEFAULT_ALPHA
) {
    // This is a window of {@link ClassificationResult}s as outputted by the {@link PoseClassifier}.
    // We run smoothing over this window of size {@link windowSize}.
    private val window: Deque<ClassificationResult>
    fun getSmoothedResult(classificationResult: ClassificationResult): ClassificationResult {
        // If we are at window size, remove the last (oldest) result.
        if (window.size == windowSize) {
            window.pollLast()
        }
        // Insert at the beginning of the window.
        window.addFirst(classificationResult)
        val allClasses: MutableSet<String> = HashSet()
        for (result in window) {
            allClasses.addAll(result.getAllClasses())
        }
        val smoothedResult = ClassificationResult()
        for (className in allClasses) {
            var factor = 1f
            var topSum = 0f
            var bottomSum = 0f
            for (result in window) {
                val value = result.getClassConfidence(className)
                topSum += factor * value
                bottomSum += factor
                factor = (factor * (1.0 - alpha)).toFloat()
            }
            smoothedResult.putClassConfidence(className, topSum / bottomSum)
        }
        return smoothedResult
    }

    companion object {
        private const val DEFAULT_WINDOW_SIZE = 10
        private const val DEFAULT_ALPHA = 0.2f
    }

    init {
        window = LinkedBlockingDeque(windowSize)
    }
}

File: Users/jeong-yeongjae/sub_tree_test/iOS/FitGAI/PoseLogic/Classification/EMASmoothing.swift
//
//  EMASmoothing.swift
//  FitGAI
//
//  Created by 김기훈 on 2021/12/13.
//

import Foundation

class EMASmoothing {
    private static let DEFAULT_WINDOW_SIZE = 10
    private static let DEFAULT_ALPHA:Float = 0.2

    let windowSize: Int = EMASmoothing.DEFAULT_WINDOW_SIZE
    let alpha: Float = EMASmoothing.DEFAULT_ALPHA
    private var window: [ClassificationResult]
    

    init() {
        window = [ClassificationResult]()
    }

    func getSmoothedResult(classificationResult: ClassificationResult) -> ClassificationResult {
        // If we are at window size, remove the last (oldest) result.
        if (window.count == windowSize) {
            window.remove(at: windowSize - 1)
        }
        // Insert at the beginning of the window.
        window.insert(classificationResult, at: 0)
        var allClasses = Set<String>()
        
        for result in window {
            for key in result.getAllClasses(){
                allClasses.insert(key)
            }
        }
        
        let smoothedResult = ClassificationResult()
        
        for className in allClasses {
            var factor:Float = 1
            var topSum:Float = 0
            var bottomSum:Float = 0
            for result in window {
                let value:Float = result.getClassConfidence(className:className)
                topSum += factor * value
                bottomSum += factor
                factor = (factor * (1.0 - alpha))
            }
            smoothedResult.putClassConfidence(className:className, confidence:topSum / bottomSum)
        }
        return smoothedResult
    }
}
